File: src/MyServiceBus/InMemoryRecurringJobProvider.cs
namespace MyServiceBus;

public sealed class InMemoryRecurringJobProvider : IRecurringJobProvider, IRecurringJobSource
{
    private sealed class Entry
    {
        public required Guid DefinitionId { get; init; }
        public required RecurringJobDefinition Definition { get; set; }
        public required object Job { get; set; }
        public required Func<CancellationToken, Task> Dispatch { get; set; }
        public required long Revision { get; set; }
        public required DateTimeOffset AcceptedAtUtc { get; set; }
        public required RecurringJobDefinitionStatus Status { get; set; }
        public DateTimeOffset? NextOccurrenceAtUtc { get; set; }
        public Guid? TimerToken { get; set; }
    }

    private readonly object gate = new();
    private readonly Dictionary<RecurringJobIdentity, Entry> definitions = [];
    private readonly HashSet<(Guid DefinitionId, long Revision, DateTimeOffset ScheduledForUtc)> occurrences = [];
    private readonly IPublishEndpoint publishEndpoint;
    private readonly ILocalDelayScheduler delayScheduler;
    private readonly TimeProvider timeProvider;

    public InMemoryRecurringJobProvider(
        IPublishEndpoint publishEndpoint,
        ILocalDelayScheduler delayScheduler,
        TimeProvider? timeProvider = null)
    {
        this.publishEndpoint = publishEndpoint;
        this.delayScheduler = delayScheduler;
        this.timeProvider = timeProvider ?? TimeProvider.System;
    }

    public string ProviderName => "InMemory";

    public SchedulingDurability Durability => SchedulingDurability.Volatile;

    public SchedulingPlacement Placement => SchedulingPlacement.ProcessLocal;

    string IRecurringJobSource.Provider => ProviderName;

    bool IRecurringJobSource.Authoritative => true;

    public Task<IReadOnlyList<RecurringJobState>> GetSnapshotAsync(
        int maximumCount,
        CancellationToken cancellationToken = default)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(maximumCount);
        cancellationToken.ThrowIfCancellationRequested();
        lock (gate)
        {
            IReadOnlyList<RecurringJobState> snapshot = definitions.Values
                .Where(entry => entry.Status != RecurringJobDefinitionStatus.Removed)
                .OrderBy(entry => entry.NextOccurrenceAtUtc ?? DateTimeOffset.MaxValue)
                .Take(maximumCount)
                .Select(entry => new RecurringJobState(
                    entry.DefinitionId,
                    entry.Definition.Identity,
                    entry.Revision,
                    ProviderName,
                    Durability,
                    Placement,
                    FormatCadence(entry.Definition.Cadence),
                    entry.Job.GetType().FullName ?? entry.Job.GetType().Name,
                    entry.Status,
                    entry.NextOccurrenceAtUtc,
                    entry.AcceptedAtUtc))
                .ToArray();
            return Task.FromResult(snapshot);
        }
    }

    public async Task<RecurringJobDefinitionReceipt> AddOrUpdate<TJob>(
        RecurringJobDefinition definition,
        TJob job,
        long? expectedRevision = null,
        CancellationToken cancellationToken = default)
        where TJob : class
    {
        ArgumentNullException.ThrowIfNull(definition);
        ArgumentNullException.ThrowIfNull(job);
        EnsureSupported(definition);

        Entry entry;
        Guid? timerToCancel;
        lock (gate)
        {
            definitions.TryGetValue(definition.Identity, out var current);
            ValidateExpectedRevision(definition.Identity, expectedRevision, current?.Revision ?? 0);

            if (current is not null
                && current.Status != RecurringJobDefinitionStatus.Removed
                && current.Definition == definition
                && Equals(current.Job, job))
            {
                return CreateReceipt(current);
            }

            timerToCancel = current?.TimerToken;
            entry = new Entry
            {
                DefinitionId = current?.DefinitionId ?? Guid.NewGuid(),
                Definition = definition,
                Job = job,
                Dispatch = token => publishEndpoint.Publish(job, cancellationToken: token),
                Revision = (current?.Revision ?? 0) + 1,
                AcceptedAtUtc = timeProvider.GetUtcNow(),
                Status = RecurringJobDefinitionStatus.Active
            };
            definitions[definition.Identity] = entry;
        }

        if (timerToCancel is Guid oldTimer)
            await delayScheduler.Cancel(oldTimer).ConfigureAwait(false);

        await ScheduleNext(entry, timeProvider.GetUtcNow(), cancellationToken).ConfigureAwait(false);
        return CreateReceipt(entry);
    }

    public async Task<RecurringJobControlResult> Pause(
        RecurringJobIdentity identity,
        long? expectedRevision = null,
        CancellationToken cancellationToken = default)
    {
        Guid? timer;
        long revision;
        lock (gate)
        {
            if (!definitions.TryGetValue(identity, out var entry)
                || entry.Status == RecurringJobDefinitionStatus.Removed)
                return new(RecurringJobControlOutcome.NotFound);

            ValidateExpectedRevision(identity, expectedRevision, entry.Revision);
            if (entry.Status == RecurringJobDefinitionStatus.Paused)
                return new(RecurringJobControlOutcome.Unchanged, entry.Revision);

            entry.Status = RecurringJobDefinitionStatus.Paused;
            entry.Revision++;
            entry.NextOccurrenceAtUtc = null;
            timer = entry.TimerToken;
            entry.TimerToken = null;
            revision = entry.Revision;
        }

        if (timer is Guid timerToken)
            await delayScheduler.Cancel(timerToken).ConfigureAwait(false);

        return new(RecurringJobControlOutcome.Applied, revision);
    }

    public async Task<RecurringJobControlResult> Resume(
        RecurringJobIdentity identity,
        long? expectedRevision = null,
        CancellationToken cancellationToken = default)
    {
        Entry entry;
        lock (gate)
        {
            if (!definitions.TryGetValue(identity, out entry!)
                || entry.Status == RecurringJobDefinitionStatus.Removed)
                return new(RecurringJobControlOutcome.NotFound);

            ValidateExpectedRevision(identity, expectedRevision, entry.Revision);
            if (entry.Status == RecurringJobDefinitionStatus.Active)
                return new(RecurringJobControlOutcome.Unchanged, entry.Revision);

            entry.Status = RecurringJobDefinitionStatus.Active;
            entry.Revision++;
        }

        await ScheduleNext(entry, timeProvider.GetUtcNow(), cancellationToken).ConfigureAwait(false);
        return new(RecurringJobControlOutcome.Applied, entry.Revision);
    }

    public async Task<RecurringJobControlResult> Remove(
        RecurringJobIdentity identity,
        long? expectedRevision = null,
        CancellationToken cancellationToken = default)
    {
        Guid? timer;
        long revision;
        lock (gate)
        {
            if (!definitions.TryGetValue(identity, out var entry)
                || entry.Status == RecurringJobDefinitionStatus.Removed)
                return new(RecurringJobControlOutcome.NotFound);

            ValidateExpectedRevision(identity, expectedRevision, entry.Revision);
            entry.Status = RecurringJobDefinitionStatus.Removed;
            entry.Revision++;
            entry.NextOccurrenceAtUtc = null;
            timer = entry.TimerToken;
            entry.TimerToken = null;
            revision = entry.Revision;
        }

        if (timer is Guid timerToken)
            await delayScheduler.Cancel(timerToken).ConfigureAwait(false);

        return new(RecurringJobControlOutcome.Applied, revision);
    }

    public async Task<RecurringJobOccurrenceReceipt> TriggerNow(
        RecurringJobIdentity identity,
        CancellationToken cancellationToken = default)
    {
        Entry entry;
        DateTimeOffset scheduledFor;
        Guid occurrenceId;
        lock (gate)
        {
            if (!definitions.TryGetValue(identity, out entry!)
                || entry.Status is RecurringJobDefinitionStatus.Removed or RecurringJobDefinitionStatus.Ended)
                throw new RecurringJobNotFoundException(identity);

            scheduledFor = timeProvider.GetUtcNow();
            occurrenceId = Guid.NewGuid();
        }

        await entry.Dispatch(cancellationToken).ConfigureAwait(false);
        return new(
            occurrenceId,
            entry.DefinitionId,
            entry.Revision,
            scheduledFor,
            true,
            RecurringJobOccurrenceStatus.Dispatched);
    }

    private async Task ScheduleNext(Entry entry, DateTimeOffset afterUtc, CancellationToken cancellationToken)
    {
        var next = CalculateNext(entry, afterUtc);
        await ScheduleAt(entry, next, cancellationToken).ConfigureAwait(false);
    }

    private async Task ScheduleAt(
        Entry entry,
        DateTimeOffset? next,
        CancellationToken cancellationToken)
    {
        if (next is null)
        {
            lock (gate)
            {
                if (IsCurrent(entry))
                {
                    entry.Status = RecurringJobDefinitionStatus.Ended;
                    entry.NextOccurrenceAtUtc = null;
                }
            }
            return;
        }

        var identity = entry.Definition.Identity;
        var revision = entry.Revision;
        var timerToken = await delayScheduler.Schedule(
            next.Value.UtcDateTime,
            token => Materialize(identity, revision, next.Value, token),
            cancellationToken).ConfigureAwait(false);

        var cancelTimer = false;
        lock (gate)
        {
            if (IsCurrent(entry) && entry.Status == RecurringJobDefinitionStatus.Active)
            {
                entry.NextOccurrenceAtUtc = next;
                entry.TimerToken = timerToken;
            }
            else
            {
                cancelTimer = true;
            }
        }

        if (cancelTimer)
            await delayScheduler.Cancel(timerToken).ConfigureAwait(false);
    }

    private async Task Materialize(
        RecurringJobIdentity identity,
        long revision,
        DateTimeOffset scheduledForUtc,
        CancellationToken cancellationToken)
    {
        Entry entry;
        int dispatchCount;
        DateTimeOffset? next;
        lock (gate)
        {
            if (!definitions.TryGetValue(identity, out entry!)
                || entry.Revision != revision
                || entry.Status != RecurringJobDefinitionStatus.Active
                || occurrences.Contains((entry.DefinitionId, revision, scheduledForUtc)))
                return;

            entry.TimerToken = null;
            entry.NextOccurrenceAtUtc = null;
            (dispatchCount, next) = EvaluateDue(entry, scheduledForUtc, timeProvider.GetUtcNow());
            occurrences.Add((entry.DefinitionId, revision, scheduledForUtc));
            for (var index = 1; index < dispatchCount; index++)
            {
                var occurrenceTime = scheduledForUtc + TimeSpan.FromTicks(
                    checked(((FixedIntervalRecurringJobCadence)entry.Definition.Cadence).Interval.Ticks * index));
                occurrences.Add((entry.DefinitionId, revision, occurrenceTime));
            }
        }

        await ScheduleAt(entry, next, cancellationToken).ConfigureAwait(false);
        for (var index = 0; index < dispatchCount; index++)
            await entry.Dispatch(cancellationToken).ConfigureAwait(false);
    }

    private bool IsCurrent(Entry entry) =>
        definitions.TryGetValue(entry.Definition.Identity, out var current)
        && ReferenceEquals(current, entry);

    private static (int DispatchCount, DateTimeOffset? Next) EvaluateDue(
        Entry entry,
        DateTimeOffset scheduledForUtc,
        DateTimeOffset nowUtc)
    {
        var following = CalculateNext(entry, scheduledForUtc);
        var isMisfire = following is { } nextAfterScheduled && nextAfterScheduled <= nowUtc;
        if (!isMisfire)
            return (1, following);

        var dispatchCount = entry.Definition.MisfirePolicy switch
        {
            RecurringJobMisfirePolicy.Skip => 0,
            RecurringJobMisfirePolicy.FireOnceNow => 1,
            RecurringJobMisfirePolicy.CatchUp => CountCatchUpOccurrences(entry, scheduledForUtc, nowUtc),
            _ => throw new NotSupportedException($"Unsupported misfire policy '{entry.Definition.MisfirePolicy}'.")
        };
        return (dispatchCount, CalculateNext(entry, nowUtc));
    }

    private static int CountCatchUpOccurrences(
        Entry entry,
        DateTimeOffset scheduledForUtc,
        DateTimeOffset nowUtc)
    {
        var cadence = (FixedIntervalRecurringJobCadence)entry.Definition.Cadence;
        var lastEligible = entry.Definition.EndAtUtc is { } end && end <= nowUtc
            ? end.AddTicks(-1)
            : nowUtc;
        var elapsedIntervals = (lastEligible - scheduledForUtc).Ticks / cadence.Interval.Ticks;
        return (int)Math.Min(checked(elapsedIntervals + 1), entry.Definition.MaxCatchUpOccurrences);
    }

    private static DateTimeOffset? CalculateNext(Entry entry, DateTimeOffset afterUtc)
    {
        var definition = entry.Definition;
        var cadence = (FixedIntervalRecurringJobCadence)definition.Cadence;
        var anchor = cadence.AnchorAtUtc ?? definition.StartAtUtc ?? entry.AcceptedAtUtc;
        var threshold = definition.StartAtUtc is { } start && start > afterUtc ? start.AddTicks(-1) : afterUtc;
        DateTimeOffset next;
        if (anchor > threshold)
        {
            next = anchor;
        }
        else
        {
            var elapsedTicks = (threshold - anchor).Ticks;
            var steps = (elapsedTicks / cadence.Interval.Ticks) + 1;
            next = anchor + TimeSpan.FromTicks(checked(steps * cadence.Interval.Ticks));
        }

        return definition.EndAtUtc is { } end && next >= end ? null : next;
    }

    private static void EnsureSupported(RecurringJobDefinition definition)
    {
        if (definition.Cadence is not FixedIntervalRecurringJobCadence)
            throw new NotSupportedException("The in-memory recurring scheduler currently supports fixed intervals only.");
        if (definition.OverlapPolicy != RecurringJobOverlapPolicy.Allow)
            throw new NotSupportedException("The dispatch-only recurring scheduler supports the Allow overlap policy only.");
    }

    private static string FormatCadence(RecurringJobCadence cadence) => cadence switch
    {
        FixedIntervalRecurringJobCadence fixedInterval => $"Every {fixedInterval.Interval}",
        CronRecurringJobCadence cron => $"{cron.Dialect}: {cron.Expression} ({cron.TimeZoneId})",
        _ => cadence.GetType().Name
    };

    private static void ValidateExpectedRevision(
        RecurringJobIdentity identity,
        long? expectedRevision,
        long currentRevision)
    {
        if (expectedRevision is { } expected && expected != currentRevision)
            throw new RecurringJobRevisionConflictException(identity, expected, currentRevision);
    }

    private RecurringJobDefinitionReceipt CreateReceipt(Entry entry) => new(
        entry.DefinitionId,
        entry.Definition.Identity,
        entry.Revision,
        ProviderName,
        Durability,
        Placement,
        entry.AcceptedAtUtc,
        entry.NextOccurrenceAtUtc);
}

File: src/Java/myservicebus/src/main/java/com/myservicebus/InMemoryRecurringJobProvider.java
package com.myservicebus;

import com.myservicebus.tasks.CancellationToken;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

public final class InMemoryRecurringJobProvider implements RecurringJobProvider, RecurringJobSource {
    private static final class Entry {
        private final UUID definitionId;
        private RecurringJobDefinition definition;
        private Object job;
        private Submission submit;
        private long revision;
        private Instant acceptedAtUtc;
        private RecurringJobDefinitionStatus status;
        private Instant nextOccurrenceAtUtc;
        private UUID timerToken;

        private Entry(
                UUID definitionId,
                RecurringJobDefinition definition,
                Object job,
                Submission submit,
                long revision,
                Instant acceptedAtUtc) {
            this.definitionId = definitionId;
            this.definition = definition;
            this.job = job;
            this.submit = submit;
            this.revision = revision;
            this.acceptedAtUtc = acceptedAtUtc;
            this.status = RecurringJobDefinitionStatus.ACTIVE;
        }
    }

    private record OccurrenceKey(UUID definitionId, long revision, Instant scheduledForUtc) {
    }

    @FunctionalInterface
    private interface Submission {
        CompletionStage<Void> apply(UUID occurrenceId, CancellationToken cancellationToken);
    }

    private final Object gate = new Object();
    private final Map<RecurringJobIdentity, Entry> definitions = new HashMap<>();
    private final Set<OccurrenceKey> occurrences = new HashSet<>();
    private final JobClient jobClient;
    private final LocalDelayScheduler delayScheduler;
    private final Clock clock;

    public InMemoryRecurringJobProvider(
            JobClient jobClient,
            LocalDelayScheduler delayScheduler) {
        this(jobClient, delayScheduler, Clock.systemUTC());
    }

    public InMemoryRecurringJobProvider(
            JobClient jobClient,
            LocalDelayScheduler delayScheduler,
            Clock clock) {
        this.jobClient = jobClient;
        this.delayScheduler = delayScheduler;
        this.clock = clock;
    }

    @Override
    public String getProviderName() {
        return "InMemory";
    }

    @Override
    public SchedulingDurability getDurability() {
        return SchedulingDurability.VOLATILE;
    }

    @Override
    public SchedulingPlacement getPlacement() {
        return SchedulingPlacement.PROCESS_LOCAL;
    }

    @Override
    public String getProvider() {
        return getProviderName();
    }

    @Override
    public boolean isAuthoritative() {
        return true;
    }

    @Override
    public CompletionStage<java.util.List<RecurringJobState>> getSnapshot(int maximumCount) {
        if (maximumCount <= 0) {
            throw new IllegalArgumentException("maximumCount must be greater than zero");
        }
        synchronized (gate) {
            return CompletableFuture.completedFuture(definitions.values().stream()
                    .filter(entry -> entry.status != RecurringJobDefinitionStatus.REMOVED)
                    .sorted(java.util.Comparator.comparing(
                            entry -> entry.nextOccurrenceAtUtc,
                            java.util.Comparator.nullsLast(java.util.Comparator.naturalOrder())))
                    .limit(maximumCount)
                    .map(entry -> new RecurringJobState(
                            entry.definitionId,
                            entry.definition.identity(),
                            entry.revision,
                            getProviderName(),
                            getDurability(),
                            getPlacement(),
                            formatCadence(entry.definition.cadence()),
                            entry.job.getClass().getName(),
                            entry.status,
                            entry.nextOccurrenceAtUtc,
                            entry.acceptedAtUtc))
                    .toList());
        }
    }

    @Override
    public <TJob> CompletionStage<RecurringJobDefinitionReceipt> addOrUpdate(
            RecurringJobDefinition definition,
            TJob job,
            Long expectedRevision,
            CancellationToken cancellationToken) {
        Objects.requireNonNull(definition, "definition");
        Objects.requireNonNull(job, "job");
        ensureSupported(definition);

        Entry entry;
        UUID timerToCancel;
        synchronized (gate) {
            Entry current = definitions.get(definition.identity());
            validateExpectedRevision(definition.identity(), expectedRevision, current == null ? 0 : current.revision);
            if (current != null
                    && current.status != RecurringJobDefinitionStatus.REMOVED
                    && current.definition.equals(definition)
                    && Objects.equals(current.job, job)) {
                return CompletableFuture.completedFuture(createReceipt(current));
            }

            timerToCancel = current == null ? null : current.timerToken;
            entry = new Entry(
                    current == null ? UUID.randomUUID() : current.definitionId,
                    definition,
                    job,
                    (occurrenceId, token) -> jobClient.submit(
                            job,
                            new JobSubmissionOptions(null, occurrenceId),
                            token)
                            .thenApply(ignored -> null),
                    (current == null ? 0 : current.revision) + 1,
                    clock.instant());
            definitions.put(definition.identity(), entry);
        }

        CompletionStage<Boolean> cancelled = timerToCancel == null
                ? CompletableFuture.completedFuture(false)
                : delayScheduler.cancel(timerToCancel);
        return cancelled.thenCompose(ignored -> scheduleNext(entry, clock.instant(), cancellationToken))
                .thenApply(ignored -> createReceipt(entry));
    }

    @Override
    public CompletionStage<RecurringJobControlResult> pause(
            RecurringJobIdentity identity,
            Long expectedRevision,
            CancellationToken cancellationToken) {
        UUID timer;
        long revision;
        synchronized (gate) {
            Entry entry = definitions.get(identity);
            if (entry == null || entry.status == RecurringJobDefinitionStatus.REMOVED) {
                return CompletableFuture.completedFuture(
                        new RecurringJobControlResult(RecurringJobControlOutcome.NOT_FOUND));
            }
            validateExpectedRevision(identity, expectedRevision, entry.revision);
            if (entry.status == RecurringJobDefinitionStatus.PAUSED) {
                return CompletableFuture.completedFuture(
                        new RecurringJobControlResult(RecurringJobControlOutcome.UNCHANGED, entry.revision));
            }
            entry.status = RecurringJobDefinitionStatus.PAUSED;
            entry.revision++;
            entry.nextOccurrenceAtUtc = null;
            timer = entry.timerToken;
            entry.timerToken = null;
            revision = entry.revision;
        }

        CompletionStage<Boolean> cancelled = timer == null
                ? CompletableFuture.completedFuture(false)
                : delayScheduler.cancel(timer);
        return cancelled.thenApply(ignored ->
                new RecurringJobControlResult(RecurringJobControlOutcome.APPLIED, revision));
    }

    @Override
    public CompletionStage<RecurringJobControlResult> resume(
            RecurringJobIdentity identity,
            Long expectedRevision,
            CancellationToken cancellationToken) {
        Entry entry;
        synchronized (gate) {
            entry = definitions.get(identity);
            if (entry == null || entry.status == RecurringJobDefinitionStatus.REMOVED) {
                return CompletableFuture.completedFuture(
                        new RecurringJobControlResult(RecurringJobControlOutcome.NOT_FOUND));
            }
            validateExpectedRevision(identity, expectedRevision, entry.revision);
            if (entry.status == RecurringJobDefinitionStatus.ACTIVE) {
                return CompletableFuture.completedFuture(
                        new RecurringJobControlResult(RecurringJobControlOutcome.UNCHANGED, entry.revision));
            }
            entry.status = RecurringJobDefinitionStatus.ACTIVE;
            entry.revision++;
        }

        return scheduleNext(entry, clock.instant(), cancellationToken).thenApply(ignored ->
                new RecurringJobControlResult(RecurringJobControlOutcome.APPLIED, entry.revision));
    }

    @Override
    public CompletionStage<RecurringJobControlResult> remove(
            RecurringJobIdentity identity,
            Long expectedRevision,
            CancellationToken cancellationToken) {
        UUID timer;
        long revision;
        synchronized (gate) {
            Entry entry = definitions.get(identity);
            if (entry == null || entry.status == RecurringJobDefinitionStatus.REMOVED) {
                return CompletableFuture.completedFuture(
                        new RecurringJobControlResult(RecurringJobControlOutcome.NOT_FOUND));
            }
            validateExpectedRevision(identity, expectedRevision, entry.revision);
            entry.status = RecurringJobDefinitionStatus.REMOVED;
            entry.revision++;
            entry.nextOccurrenceAtUtc = null;
            timer = entry.timerToken;
            entry.timerToken = null;
            revision = entry.revision;
        }

        CompletionStage<Boolean> cancelled = timer == null
                ? CompletableFuture.completedFuture(false)
                : delayScheduler.cancel(timer);
        return cancelled.thenApply(ignored ->
                new RecurringJobControlResult(RecurringJobControlOutcome.APPLIED, revision));
    }

    @Override
    public CompletionStage<RecurringJobOccurrenceReceipt> triggerNow(
            RecurringJobIdentity identity,
            CancellationToken cancellationToken) {
        Entry entry;
        Instant scheduledFor;
        UUID occurrenceId = UUID.randomUUID();
        synchronized (gate) {
            entry = definitions.get(identity);
            if (entry == null
                    || entry.status == RecurringJobDefinitionStatus.REMOVED
                    || entry.status == RecurringJobDefinitionStatus.ENDED) {
                throw new RecurringJobNotFoundException(identity);
            }
            scheduledFor = clock.instant();
        }

        Entry captured = entry;
        return entry.submit.apply(occurrenceId, cancellationToken).thenApply(ignored ->
                new RecurringJobOccurrenceReceipt(
                        occurrenceId,
                        captured.definitionId,
                        captured.revision,
                        scheduledFor,
                        true,
                        RecurringJobOccurrenceStatus.PENDING));
    }

    private CompletionStage<Void> scheduleNext(
            Entry entry,
            Instant afterUtc,
            CancellationToken cancellationToken) {
        Instant next = calculateNext(entry, afterUtc);
        return scheduleAt(entry, next, cancellationToken);
    }

    private CompletionStage<Void> scheduleAt(
            Entry entry,
            Instant next,
            CancellationToken cancellationToken) {
        if (next == null) {
            synchronized (gate) {
                if (isCurrent(entry)) {
                    entry.status = RecurringJobDefinitionStatus.ENDED;
                    entry.nextOccurrenceAtUtc = null;
                }
            }
            return CompletableFuture.completedFuture(null);
        }

        RecurringJobIdentity identity = entry.definition.identity();
        long revision = entry.revision;
        return delayScheduler.schedule(
                next,
                token -> materialize(identity, revision, next, token),
                cancellationToken).thenCompose(timerToken -> {
                    boolean cancelTimer;
                    synchronized (gate) {
                        cancelTimer = !isCurrent(entry)
                                || entry.status != RecurringJobDefinitionStatus.ACTIVE;
                        if (!cancelTimer) {
                            entry.nextOccurrenceAtUtc = next;
                            entry.timerToken = timerToken;
                        }
                    }
                    if (cancelTimer) {
                        return delayScheduler.cancel(timerToken).thenApply(ignored -> null);
                    }
                    return CompletableFuture.completedFuture(null);
                });
    }

    private CompletionStage<Void> materialize(
            RecurringJobIdentity identity,
            long revision,
            Instant scheduledForUtc,
            CancellationToken cancellationToken) {
        Entry entry;
        int dispatchCount;
        Instant next;
        synchronized (gate) {
            entry = definitions.get(identity);
            if (entry == null
                    || entry.revision != revision
                    || entry.status != RecurringJobDefinitionStatus.ACTIVE
                    || occurrences.contains(new OccurrenceKey(entry.definitionId, revision, scheduledForUtc))) {
                return CompletableFuture.completedFuture(null);
            }
            entry.timerToken = null;
            entry.nextOccurrenceAtUtc = null;
            Evaluation evaluation = evaluateDue(entry, scheduledForUtc, clock.instant());
            dispatchCount = evaluation.dispatchCount();
            next = evaluation.next();
            occurrences.add(new OccurrenceKey(entry.definitionId, revision, scheduledForUtc));
            Duration interval = ((FixedIntervalRecurringJobCadence) entry.definition.cadence()).interval();
            for (int index = 1; index < dispatchCount; index++) {
                occurrences.add(new OccurrenceKey(
                        entry.definitionId,
                        revision,
                        scheduledForUtc.plus(interval.multipliedBy(index))));
            }
        }

        Entry captured = entry;
        CompletionStage<Void> dispatches = scheduleAt(entry, next, cancellationToken);
        for (int index = 0; index < dispatchCount; index++) {
            dispatches = dispatches.thenCompose(ignored ->
                    captured.submit.apply(UUID.randomUUID(), cancellationToken));
        }
        return dispatches;
    }

    private boolean isCurrent(Entry entry) {
        return definitions.get(entry.definition.identity()) == entry;
    }

    private record Evaluation(int dispatchCount, Instant next) {
    }

    private static Evaluation evaluateDue(Entry entry, Instant scheduledForUtc, Instant nowUtc) {
        Instant following = calculateNext(entry, scheduledForUtc);
        boolean misfire = following != null && !following.isAfter(nowUtc);
        if (!misfire) {
            return new Evaluation(1, following);
        }

        int dispatchCount = switch (entry.definition.misfirePolicy()) {
            case SKIP -> 0;
            case FIRE_ONCE_NOW -> 1;
            case CATCH_UP -> countCatchUpOccurrences(entry, scheduledForUtc, nowUtc);
        };
        return new Evaluation(dispatchCount, calculateNext(entry, nowUtc));
    }

    private static int countCatchUpOccurrences(Entry entry, Instant scheduledForUtc, Instant nowUtc) {
        Duration interval = ((FixedIntervalRecurringJobCadence) entry.definition.cadence()).interval();
        Instant lastEligible = entry.definition.endAtUtc() != null
                && !entry.definition.endAtUtc().isAfter(nowUtc)
                ? entry.definition.endAtUtc().minusNanos(1)
                : nowUtc;
        long elapsedIntervals = Duration.between(scheduledForUtc, lastEligible).dividedBy(interval);
        return (int) Math.min(Math.addExact(elapsedIntervals, 1), entry.definition.maxCatchUpOccurrences());
    }

    private static Instant calculateNext(Entry entry, Instant afterUtc) {
        RecurringJobDefinition definition = entry.definition;
        FixedIntervalRecurringJobCadence cadence =
                (FixedIntervalRecurringJobCadence) definition.cadence();
        Instant anchor = cadence.anchorAtUtc() != null
                ? cadence.anchorAtUtc()
                : definition.startAtUtc() != null ? definition.startAtUtc() : entry.acceptedAtUtc;
        Instant threshold = definition.startAtUtc() != null && definition.startAtUtc().isAfter(afterUtc)
                ? definition.startAtUtc().minusNanos(1)
                : afterUtc;
        Instant next;
        if (anchor.isAfter(threshold)) {
            next = anchor;
        } else {
            Duration elapsed = Duration.between(anchor, threshold);
            long steps = elapsed.dividedBy(cadence.interval()) + 1;
            next = anchor.plus(cadence.interval().multipliedBy(steps));
        }
        return definition.endAtUtc() != null && !next.isBefore(definition.endAtUtc()) ? null : next;
    }

    private static void ensureSupported(RecurringJobDefinition definition) {
        if (!(definition.cadence() instanceof FixedIntervalRecurringJobCadence)) {
            throw new UnsupportedOperationException(
                    "The in-memory recurring scheduler currently supports fixed intervals only.");
        }
        if (definition.overlapPolicy() != RecurringJobOverlapPolicy.ALLOW) {
            throw new UnsupportedOperationException(
                    "The dispatch-only recurring scheduler supports the ALLOW overlap policy only.");
        }
    }

    private static String formatCadence(RecurringJobCadence cadence) {
        if (cadence instanceof FixedIntervalRecurringJobCadence fixedInterval) {
            return "Every " + fixedInterval.interval();
        }
        if (cadence instanceof CronRecurringJobCadence cron) {
            return cron.dialect() + ": " + cron.expression() + " (" + cron.timeZoneId() + ")";
        }
        return cadence.getClass().getSimpleName();
    }

    private static void validateExpectedRevision(
            RecurringJobIdentity identity,
            Long expectedRevision,
            long currentRevision) {
        if (expectedRevision != null && expectedRevision != currentRevision) {
            throw new RecurringJobRevisionConflictException(identity, expectedRevision, currentRevision);
        }
    }

    private RecurringJobDefinitionReceipt createReceipt(Entry entry) {
        return new RecurringJobDefinitionReceipt(
                entry.definitionId,
                entry.definition.identity(),
                entry.revision,
                getProviderName(),
                getDurability(),
                getPlacement(),
                entry.acceptedAtUtc,
                entry.nextOccurrenceAtUtc);
    }
}
